File: backend/src/main/java/com/proj/sms/service/SubjectService.java
package com.proj.sms.service;

import com.proj.sms.interfaces.SubjectRepository;
import com.proj.sms.models.Subject;
import org.springframework.stereotype.Service;
import java.util.List;

@Service
public class SubjectService {

    private final SubjectRepository subjectRepository;

    public SubjectService(SubjectRepository subjectRepository) {
        this.subjectRepository = subjectRepository;
    }

    // CREATE
    public Subject createSubject(Subject subject) {
        return subjectRepository.save(subject);
    }

    // READ (Single)
    public Subject getSubjectById(Long id) {
        return subjectRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Subject not found with id: " + id));
    }

    // READ (All)
    public List<Subject> getAllSubjects() {
        return subjectRepository.findAll();
    }

    // ✅ ADD THIS METHOD
    // READ (Search by Name)
    public List<Subject> searchSubjectsByName(String keyword) {
        return subjectRepository.findByNameContainingIgnoreCase(keyword);
    }

    // UPDATE
    public Subject updateSubject(Long id, Subject subjectDetails) {
        Subject subject = getSubjectById(id);
        subject.setName(subjectDetails.getName());
        return subjectRepository.save(subject);
    }

    // DELETE
    public void deleteSubject(Long id) {
        Subject subject = getSubjectById(id);
        subjectRepository.delete(subject);
    }

    public boolean existsByName(String name) {
        return subjectRepository.existsByName(name);
    }
}

File: frontend/src/services/subjectService.js
// src/services/subjectService.js
import api from './api';

const SUBJECTS_API_URL = '/subjects';

export const getAllSubjects = async (keyword = '') => {
    const response = await api.get(SUBJECTS_API_URL, {
        params: { keyword: keyword } // Pass keyword as a query parameter
    });
    return response.data;
};

export const getSubjectById = async (id) => {
    const response = await api.get(`${SUBJECTS_API_URL}/${id}`);
    return response.data;
};

export const createSubject = async (subject) => {
    const response = await api.post(SUBJECTS_API_URL, subject);
    return response.data;
};

export const updateSubject = async (id, subjectDetails) => {
    const response = await api.put(`${SUBJECTS_API_URL}/${id}`, subjectDetails);
    return response.data;
};

export const deleteSubject = async (id) => {
    await api.delete(`${SUBJECTS_API_URL}/${id}`);
};